### PR TITLE
disttask: skip scheduler take slots for some states

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #49008 

Problem Summary:
1. For reverting/cancelling/pausing, we can start scheduler without allocating slots
2. As scheduler is a loop. For a task in pausing state, if the scheduler is started without allocating slots, and if the task enters the running state by another scheduler, or enters the resuming state in the refresh interval, it should not start running without allocating slots(just exit).
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/disttask/framework/scheduler/scheduler_manager_nokit_test.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		proto.TaskStateSucceed).Return(tasks, nil)
+	taskMgr.EXPECT().TransferTasks2History(mgr.ctx, tasks).Return(nil)
+	mgr.doCleanupTask()
+	require.True(t, ctrl.Satisfied())
+
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/pkg/disttask/framework/scheduler/WaitCleanUpFinished"))
+}
+
+func TestManagerSchedulerNotAllocateSlots(t *testing.T) {
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/disttask/framework/scheduler/exitScheduler", "return()"))
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply pkg/disttask/framework/scheduler/scheduler_manager_nokit_test.go:115]: I use manager to test allocateSlots for each scheduler.

[the remark this answers is not in the document]
Comment (ywqzzy) [reply pkg/disttask/framework/scheduler/scheduler_manager_nokit_test.go:116]: Will refine it next week

File: pkg/disttask/framework/scheduler/scheduler_nokit_test.go [hunk 2]
```diff
@@ -219,51 +186,75 @@ func TestSchedulerIsStepSucceed(t *testing.T) {
...
-		proto.TaskStateFailed,
-		proto.TaskStateReverted,
-		proto.TaskStateSucceed).Return(tasks, nil)
+	cloneTask := task
+	sch := createScheduler(&cloneTask, false, taskMgr, ctrl)
+	taskMgr.EXPECT().GetTaskByID(gomock.Any(), cloneTask.ID).DoAndReturn(func(_ context.Context, _ int64) (*proto.Task, error) {
+		cloneTask.State = proto.TaskStateResuming
+		return &cloneTask, nil
+	})
+	sch.scheduleTask()
```
Comment (D3Hunter) [inline pkg/disttask/framework/scheduler/scheduler_nokit_test.go:209]:
```suggestion
	sch.scheduleTask()
	require.True(t, ctrl.Satisfied())
```
below too

**Review by D3Hunter — APPROVED** [review]
rest lgtm

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**